libraries/chain/include/scorum/chain/database/budget_management_algorithms.hpp

```diff
@@ -0,0 +1,252 @@
+        auto head_block_time = _dgp_service.head_block_time();
+
+        auto advance = (start_date.sec_since_epoch() - head_block_time.sec_since_epoch()) / SCORUM_BLOCK_INTERVAL;
+        // budget must be allocated exactly from start time but doesn't for fund budget (when head_block_num is 0)
+        auto last_cashout_block = head_block_num;
+        if (last_cashout_block + advance > 0)
+        {
+            last_cashout_block += advance - 1;
+        }
+
+        return _budget_service.create([&](object_type& budget) {
+            budget.owner = owner;
+            if (!content_permlink.empty())
+            {
+                fc::from_string(budget.content_permlink, content_permlink);
+            }
+            budget.created = head_block_time;
+            budget.start = start_date;
+            budget.deadline = end_date;
+            budget.balance = balance;
+            budget.per_block = per_block;
+            budget.last_cashout_block = last_cashout_block;
+        });
+    }
+
+    asset allocate_cash(const object_type& budget)
+    {
+        auto head_block_num = _dgp_service.head_block_num();
+
+        if (budget.last_cashout_block >= head_block_num)
+        {
+            return asset(0, budget.balance.symbol()); // empty (allocation waits new block)
+        }
+
+        FC_ASSERT(budget.per_block.amount > 0, "Invalid per_block.");
+        asset ret = decrease_balance(budget, budget.per_block);
+
+        if (!check_close_conditions(budget))
+        {
+            _budget_service.update(budget, [&](object_type& b) { b.last_cashout_block = head_block_num; });
+        }
+        else
+        {
+            close_budget_internal(budget);
+        }
+
+        return ret;
+    }
+
+protected:
+    asset calculate_per_block(const time_point_sec& start_date, const time_point_sec& end_date, const asset& balance)
+    {
+        FC_ASSERT(start_date.sec_since_epoch() < end_date.sec_since_epoch(),
+                  "Invalid date interval. Start time ${1} must be less end time ${2}",
+                  ("1", start_date)("2", end_date));
+
+        auto ret = balance;
+
+        // calculate time interval in seconds.
+        // SCORUM_BLOCK_INTERVAL must be in seconds!
+        uint32_t delta_in_sec = end_date.sec_since_epoch() - start_date.sec_since_epoch();
+
+        // multiply before division to prevent integer clipping to zero
+        ret *= SCORUM_BLOCK_INTERVAL;
+        ret /= delta_in_sec;
+
+        // non zero budget must return at least one satoshi
+        if (ret.amount < 1)
+        {
+            ret.amount = 1;
+        }
+
+        return ret;
+    }
+
+    asset decrease_balance(const object_type& budget, const asset& balance)
+    {
+        FC_ASSERT(balance.amount > 0, "Invalid balance.");
+
+        asset ret(0, balance.symbol());
+
+        _budget_service.update(budget, [&](object_type& b) {
+            if (b.balance.amount > 0 && balance.amount <= b.balance.amount)
```

Comment:
This code can be simplified: `ret = std::min(b.balance, balance); b.balance -= ret;`